File: DeskHelp.cpp
#include <iostream>
#include <vector>
#include <string>
#include <iomanip>
#include <algorithm>
#include <ctime>

using namespace std;

// Estrutura de dados para o Chamado
struct Chamado {
    int numero;
    string usuario;
    string setor;
    string descricao;
    int prioridade; // 1-Baixa, 2-Média, 3-Alta
    char status;    // A-Aberto, E-Em Atendimento, F-Fechado
};

// Variáveis Globais para controle
vector<Chamado> listaChamados;
int proximoSequencial = 1;

// Função para gerar o número de 7 dígitos (5 aleatórios + 2 sequenciais)
int gerarNumeroChamado(int seq) {
    int prefixo = (rand() % 90000) + 10000; // Gera entre 10000 e 99999
    string s_prefixo = to_string(prefixo);
    
    // Formata o sufixo para ter 2 dígitos (ex: 01, 02)
    string s_sufixo = (seq < 10) ? "0" + to_string(seq) : to_string(seq);
    
    return stoi(s_prefixo + s_sufixo);
}

// Função de comparação para ordenar a lista
// Prioridade alta primeiro, depois por ordem de chegada (sequencial)
bool compararChamados(const Chamado &a, const Chamado &b) {
    if (a.prioridade != b.prioridade) {
        return a.prioridade > b.prioridade; // Decrescente (3 > 2 > 1)
    }
    return (a.numero % 100) < (b.numero % 100); // Crescente pelos últimos 2 dígitos
}

void abrirChamado() {
    Chamado novo;
    novo.numero = gerarNumeroChamado(proximoSequencial);
    
    cout << "\n[Abertura] Numero do chamado gerado: " << novo.numero << endl;
    cout << "Nome do usuario: ";
    cin.ignore();
    getline(cin, novo.usuario);
    cout << "Setor: ";
    getline(cin, novo.setor);
    cout << "Descricao: ";
    getline(cin, novo.descricao);
    
    novo.prioridade = 1; // Padrao Baixa
    novo.status = 'A';   // Padrao Aberto
    
    listaChamados.push_back(novo);
    proximoSequencial++;
    sort(listaChamados.begin(), listaChamados.end(), compararChamados);
    cout << "Chamado aberto com sucesso!\n";
}

void localizarChamado() {
    int num;
    cout << "\nDigite o numero do chamado: ";
    cin >> num;
    
    for (auto &c : listaChamados) {
        if (c.numero == num) {
            cout << "\n--- Detalhes do Chamado ---\n";
            cout << "ID: " << c.numero << " | Status: " << c.status << endl;
            cout << "Usuario: " << c.usuario << " | Setor: " << c.setor << endl;
            cout << "Problema: " << c.descricao << endl;
            
            cout << "\nOpcoes: [A] Atualizar Prioridade | [E] Encerrar | [S] Sair: ";
            char op;
            cin >> op;
            op = toupper(op);
            
            if (op == 'A') {
                cout << "Nova Prioridade (1-Baixa, 2-Media, 3-Alta): ";
                cin >> c.prioridade;
                sort(listaChamados.begin(), listaChamados.end(), compararChamados);
                cout << "Prioridade atualizada!\n";
            } else if (op == 'E') {
                c.status = 'F';
                cout << "Chamado encerrado (Status F).\n";
            }
            return;
        }
    }
    cout << "Chamado nao encontrado.\n";
}

void excluirChamado() {
    int num;
    cout << "\nDigite o numero para EXCLUIR: ";
    cin >> num;
    
    for (auto it = listaChamados.begin(); it != listaChamados.end(); ++it) {
        if (it->numero == num) {
            listaChamados.erase(it);
            cout << "Chamado removido da lista.\n";
            return;
        }
    }
    cout << "Chamado nao encontrado.\n";
}

void imprimirTodos() {
    if (listaChamados.empty()) {
        cout << "\nLista vazia.\n";
        return;
    }
    
    cout << "\n" << setfill('=') << setw(70) << "" << endl;
    cout << setfill(' ') << left << setw(5) << "SEQ" << " | " 
         << setw(10) << "ID TOTAL" << " | " 
         << setw(15) << "USUARIO" << " | " 
         << setw(10) << "PRIO" << " | " << "STATUS" << endl;
    cout << setw(70) << setfill('-') << "" << setfill(' ') << endl;

    for (const auto &c : listaChamados) {
        string seq = to_string(c.numero).substr(5, 2);
        string p_nome = (c.prioridade == 3) ? "Alta" : (c.prioridade == 2 ? "Media" : "Baixa");
        
        cout << left << setw(5) << seq << " | " 
             << setw(10) << c.numero << " | " 
             << setw(15) << c.usuario << " | " 
             << setw(10) << p_nome << " | " << c.status << endl;
    }
}

int main() {
    srand(time(0)); // Inicializa semente aleatoria
    int opcao;
    
    do {
        cout << "\n--- HELP DESK IFNMG (C++) ---\n";
        cout << "1 - Abrir chamado\n";
        cout << "2 - Localizar chamado\n";
        cout << "3 - Excluir chamado\n";
        cout << "4 - Quantidade de chamados\n";
        cout << "5 - Imprimir chamados\n";
        cout << "6 - Sair\n";
        cout << "Escolha: ";
        cin >> opcao;
        
        switch(opcao) {
            case 1: abrirChamado(); break;
            case 2: localizarChamado(); break;
            case 3: excluirChamado(); break;
            case 4: cout << "\nTotal: " << listaChamados.size() << endl; break;
            case 5: imprimirTodos(); break;
        }
    } while (opcao != 6);
    
    return 0;
}

File: DeskHelp.py
import random

chamados = []
proximo_sequencial = 1

def gerar_numero_chamado(seq):
    # Gera 5 dígitos aleatórios + 2 dígitos sequenciais
    prefixo = random.randint(10000, 99999)
    sufixo = str(seq).zfill(2)
    return int(f"{prefixo}{sufixo}")

def organizar_lista():
    """Ordena por Prioridade (3->1) e depois por Sequencial (01->99)."""
    global chamados
    chamados.sort(key=lambda c: (-c['prioridade'], c['numero'] % 100))

def abrir_chamado():
    global proximo_sequencial
    numero = gerar_numero_chamado(proximo_sequencial)
    
    print(f"\n[Abertura] Nº do chamado em tela: {numero}")
    nome = input("Nome do usuário: ")
    setor = input("Setor do usuário: ")
    descricao = input("Descrição do problema: ")
    
    # Regra: Prioridade 1 (Baixa) e Status A (Aberto) por padrão
    novo_chamado = {
        "numero": numero,
        "usuario": nome,
        "setor": setor,
        "descricao": descricao,
        "prioridade": 1,
        "status": "A"
    }
    
    chamados.append(novo_chamado)
    proximo_sequencial += 1
    organizar_lista()
    print("Chamado aberto com sucesso!")

def localizar_chamado():
    try:
        num = int(input("\nDigite o nº do chamado para localizar: "))
        for c in chamados:
            if c["numero"] == num:
                exibir_detalhes(c)
                
                # Sub-menu para Encerrar ou Atualizar Prioridade
                print("\nOpções para este chamado:")
                print("A - Atualizar Prioridade | E - Encerrar Chamado | S - Sair")
                op = input("Escolha: ").upper()
                
                if op == 'A':
                    nova = int(input("Nova Prioridade (1-Baixa, 2-Média, 3-Alta): "))
                    if nova in [1, 2, 3]:
                        c["prioridade"] = nova
                        organizar_lista()
                        print("Prioridade atualizada!")
                elif op == 'E':
                    c["status"] = "F" # F para Fechado
                    print("Status alterado para Fechado!")
                return
        print("Chamado não encontrado.")
    except ValueError:
        print("Entrada inválida.")

def excluir_chamado():
    try:
        num = int(input("\nDigite o nº do chamado para EXCLUIR: "))
        for i, c in enumerate(chamados):
            if c["numero"] == num:
                del chamados[i] # Elimina o chamado da lista
                print(f"Chamado {num} excluído permanentemente.")
                return
        print("Chamado não encontrado.")
    except ValueError:
        print("Entrada inválida.")

def imprimir_todos():
    if not chamados:
        print("\nNenhum chamado ativo.")
        return
    
    organizar_lista()
    print("\n" + "="*75)
    print(f"{'SEQ':<4} | {'ID TOTAL':<10} | {'USUÁRIO':<15} | {'PRIORIDADE':<10} | {'STATUS':<10}")
    print("-" * 75)
    
    prio_nomes = {1: "Baixa", 2: "Média", 3: "Alta"}
    status_nomes = {"A": "Aberto", "E": "Em Atend.", "F": "Fechado"}
    
    for c in chamados:
        seq_visual = str(c['numero'])[-2:] # Pega apenas os 2 últimos dígitos
        print(f"{seq_visual:<4} | {c['numero']:<10} | {c['usuario']:<15} | "
              f"{prio_nomes[c['prioridade']]:<10} | {status_nomes[c['status']]:<10}")
    print("="*75)

def exibir_detalhes(c):
    prio_nomes = {1: "Baixa", 2: "Média", 3: "Alta"}
    status_nomes = {"A": "Aberto", "E": "Em Atendimento", "F": "Fechado"}
    print(f"\n--- Detalhes: {c['numero']} ---")
    print(f"Usuário: {c['usuario']} | Setor: {c['setor']}")
    print(f"Descrição: {c['descricao']}")
    print(f"Prioridade: {prio_nomes[c['prioridade']]} | Status: {status_nomes[c['status']]}")

def menu():
    while True:
        print("\n--- MENU HELP DESK ---")
        print("1 - Abrir chamado")
        print("2 - Localizar chamado (por nº)")
        print("3 - Excluir chamado")
        print("4 - Quantidade de chamados (todos)")
        print("5 - Imprimir chamados (todos)")
        print("6 - Sair do sistema")
        
        opcao = input("\nEscolha: ")
        
        if opcao == '1': abrir_chamado()
        elif opcao == '2': localizar_chamado()
        elif opcao == '3': excluir_chamado()
        elif opcao == '4': print(f"\nTotal: {len(chamados)} chamados.")
        elif opcao == '5': imprimir_todos()
        elif opcao == '6': break
        else: print("Opção inválida.")

if __name__ == "__main__":
    menu()